torch/testing/_internal/common_device_type.py

```diff
@@ -916,7 +903,30 @@ def instantiate_device_type_tests(
 
         # type set to Any and suppressed due to unsupport runtime class:
         # https://github.com/python/mypy/wiki/Unsupported-Python-Features
-        device_type_test_class: Any = type(class_name, (base, empty_class), {})
+        device_type_test_class: Any = type(class_name, (base, generic_test_class), {})
+
+        # Arrange for setUpClass and tearDownClass methods defined both in the test template
+        # class and in the generic base to be called. This allows device-parameterized test
+        # classes to support setup and teardown.
+        # NB: This should be done before instantiate_test() is called as that invokes setup.
+        @classmethod
+        def _setUpClass(cls):
+            # This should always be called, whether or not the test class invokes
+            # super().setUpClass(), to set the primary device.
+            base.setUpClass()
+            # We want to call the @classmethod defined in the generic base, but pass
+            # it the device-specific class object (cls), hence the __func__ call.
+            generic_test_class.setUpClass.__func__(cls)
```

Comment:
so are we able to call super here now?

Reply:
not quite, this method conceptually is split into three parts:
1. Setup for `{CPU, CUDA}TestBase`; this sets the primary device, for example. This should always be called!
2. Setup for superclasses of `TestFoo`. This is optionally called, as defined by the author of `TestFoo`; they should make the `super()` call if they want this behavior.
3. Setup for `TestFoo` itself.

if we just call `super()` here, then we'll always do all three. But we want (2) to be optional as decided by the test case author; they decide whether or not to make the `super()` call in `TestFoo.setUpClass()`.